### ISSUE TITLE
Consistent CPU usage (and high-ish memory usage?)

### ISSUE THREAD

**Issue by jstrese** [emesene/emesene#1238]
**Version:** 2.12.3
**OS:** Windows 7 x64
**Screenshot:** http://imgur.com/AIjKM

**Custom Settings**
**Theme:** Radiance
**Plugins:** Idle status, Typing notification 

Emesene seems to be holding out at 12-13% CPU usage. I have no Emesene windows active, so it's weird. It shouldn't be doing much of anything. I've had Emesene running for about 48hrs now.

It's also using 134,000 K of memory (as seen in the screenshot). Is this normal? It seems rather high.

I restarted Emesene and the CPU usage went down to 0 as expected. I suspect that it's getting stuck somewhere :(

**Comment by pepeleproso** [comment]
We have memory leaks but we are unable to find where they are. Some month ago sbte spend quite some time debugging that but we can't find a proper solution.

**Comment by jstrese** [comment]
If there's any way I can help with that let me know.

**Comment by Sbte** [comment]
It's probably some timeouts. Maybe we can check for those.

**Comment by lfottaviano** [comment]
Its useful to us to know, what you were doing with emesene. If you say that you left emesene connected 48 hours, when almost all time you were away, we can think that there could be problems in the contact list related things. (Users disconnect/connect events and related things).  I think should be useful also, to know what protocol were you using, no?.

**Comment by jstrese** [comment]
I was using the MSN protocol. I was AFK about a third of that time (not at one time).

If it helps at all: I get "Message could not be sent" quite often, but the recipient seems to always get what I send so I think it's a flaw in the application :/

If you want me to run another version with more internal logging/debugging set up let me know.

**Comment by c10ud** [comment]
I don't know what's the typical ram usage for emesene in w7_x64, i get ~60Mb in ubuntu just opening the application, the extra ram could be python not doing proper gc, i remember in the past we fixed huge leaks. since also cpu is involved here that could be some huge animated avatar (?)

as for the message could not be sent, it's #1171

**Comment by jstrese** [comment]
I've had a new instance running since I posted this and it's only up to 67Mb~

**Comment by c10ud** [comment]
so, @lfottaviano found that we're leaking quite a lot when doing the disconnected-reconnect thing:
we need to hack dummy backend for simulating this (easy, just fire the disconnected event with a timeout just like we're doing with contact-added-you windows...which could be removed or commented out from dummy, btw) and then see if we can free that memory efficiently

**Comment by Sbte** [comment]
What we need is a dummy event firing panel. So we can fire events like contact added you, disconnection, that sort of stuff. And a FIRE THEM ALL button that we can use to test ALL THE THINGS.

**Comment by c10ud** [comment]
nah, clicking is too much work, it's better haxing manually everytime we want to test something -- in this case i hear it's disconnection :)

**Comment by Sbte** [comment]
We have had to test disconnection stuff many times already. Time for a button :P

**Comment by c10ud** [comment]
see gui/gtkui/UserPanel.py and search for "do_something_weird"

2012/5/25 Sven reply@reply.github.com:

**Comment by Sbte** [comment]
I know, I know. But it'd be easier to have a button for that that's always there...

**Comment by c10ud** [comment]
@pepeleproso changes for #444 affected this too (?)

**Comment by c10ud** [comment]
@pepeleproso @Sbte did we solve this with the extension work?

**Comment by pepeleproso** [comment]
The situation is improved but the mainwindow is no property disposed yet.

**Comment by Sbte** [comment]
@c10ud, I don't think this will ever be really fixed. I fixed the two most obvious bugs, which should help, and Jose fixed a lot more, so it should help, but I'm not sure if it's enough.

Also, we didn't do a lot about memory usage yet.

**Comment by c10ud** [comment]
well, if we don't leak that much, it's an improvement to me

i asked because i wasn't sure leaks with mainwindow-loginwindow
switching were completely fixed or not or partially or whatever

2012/6/13 Sven reply@reply.github.com:

**Comment by c10ud** [comment]
conversation leaks could be fixed by 1656f829489bbf93ca7e80e25681c7f09a592da4

**Comment by Sbte** [comment]
@c10ud that only made it worse...

**Comment by c10ud** [comment]
dict                           4286      +168
instancemethod                  372      +109
list                           1669      +103
tuple                          2475       +48
function                      10024       +40
builtin_function_or_method     2725       +38
Contact                          66       +33
Account                          64       +32
ImageMenuItem                    60       +30
_Condition                       38       +18
Image                            26       +13
Button                           22       +11
frame                            30       +10
cell                            179       +10
weakref                        2276       +10
wrapper_descriptor             2095       +10
Group                            14        +7
Menu                             12        +6
CheckButton                      12        +6
Label                            10        +5
instance                         64        +5
CheckMenuItem                    10        +5
Entry                            10        +5
StatusMenu                        8        +4
deque                             9        +4
Signal                           10        +4
Avatar                            8        +4
HBox                              8        +4
LogRecord                         6        +3
ContactHandler                    6        +3
_Event                            7        +3
Pixbuf                           30        +3
StatusButton                      6        +3
ComboBoxEntry                     4        +2
SmileyLayout                      4        +2
TextField                         4        +2
ContactMenu                       4        +2
member_descriptor               358        +2
ComboBox                          4        +2
ListStore                         4        +2
RadioMenuItem                     4        +2
TreeViewColumn                    4        +2
getset_descriptor              1978        +2
SmileyLabel                       4        +2
NiceBar                           4        +2
Layout                            4        +2
OptionsMenu                       2        +1
EventBox                          2        +1
MainMenu                          2        +1
Signals                           2        +1
TreeModelFilter                   2        +1
OptionsHandler                    2        +1
Config                            3        +1
LoggerProcess                     2        +1
Tooltips                          2        +1
LoggerOutputProcess               2        +1
WeakMethodBound                   5        +1
method_descriptor              6108        +1
CellRendererPlus                  2        +1
EndPointsMenu                     2        +1
MainWindow                        2        +1
FileHandler                       2        +1
Proxy                             2        +1
ContactManager                    2        +1
ToggleButton                      2        +1
GroupHandler                      2        +1
Logger                          109        +1
AvatarRenderer                    2        +1
ConnectingWindow                  2        +1
MyAccountHandler                  2        +1
Login                             2        +1
HelpHandler                       2        +1
ConfigDir                         3        +1
UserPanel                         2        +1
AccelGroup                        3        +1
FileMenu                          2        +1
Session                           2        +1
ContactList                       2        +1
MenuItem                          2        +1
TreeStore                         2        +1
Event                             2        +1

**Comment by c10ud** [comment]
Gtk signals are keeping references everywhere, as talked with pepeleproso we must find a nice way to weak-connect them :)

@sbte @marianoguerra

otherwise we should disconnect every signal by hand, not sure on what's the best thing to do here

**Comment by c10ud** [comment]
this weak_connect thing looks effective reading objgraph's output -- however i'm still unsure on how to handle signals which are connected to lambda methods because they don't work with this implementation out of the box :)

@pepeleproso

**Comment by Sbte** [comment]
@c10ud maybe we need a weakref to the function in WeakMethodBound too? like:

```
class WeakMethodBound(object):

    def __init__(self, f):
        self.f = weakref.ref(f.im_func)
        self.c = weakref.ref(f.im_self)

    def __call__(self, *arg):
        obj = self.f()
        fun = self.c()
        if fun is None:
            raise TypeError('Method called on dead object')

        fun(obj, *arg)
```

**Comment by Sbte** [comment]
Also, when you have that, you can derive CallbackWrapper from WeakMethod. In that case it will also work for non-class methods. Which is nicer.

**Comment by c10ud** [comment]
i'm a bit lost here, i need a hand from the python gods

2012/7/1 Sven reply@reply.github.com:

**Comment by Sbte** [comment]
7de8ef87fc2e74fea21cb835ae64102c0b9ab828 fixes another leak

**Comment by c10ud** [comment]
sbte, care to rerun the test now that you destroy the window?
1. enable the autodisconnect in dummy session
2. enable the objgraph code in emesene.py
3. start emesene and set dummy session as autologin
4. watch in debug references growing
5. ???
